src/osd/OSD.h

```diff
@@ -529,54 +529,76 @@ class OSDService {
    * down, without worrying about reopening connections from threads
    * working from old maps.
    */
+  using OSDMapRaw = const OSDMap*;
 private:
-  OSDMapRef next_osdmap;
+  // Uses two element sized vector to store two OSDMap references. Since
+  // we must await old osdmap release, so up to 2 element size is enough.
+  std::vector<OSDMapRef> saved_osdmaps;
+  std::atomic<OSDMapRaw> next_osdmap;
   Cond pre_publish_cond;
 
 public:
   void pre_publish_map(OSDMapRef map) {
     Mutex::Locker l(pre_publish_lock);
-    next_osdmap = std::move(map);
+    if (next_osdmap)
+      saved_osdmaps[1] = map;
```

Comment:
next_osdmap will essentially always be non-null (except just after boot).  I don't really get what's happening here.